### PR TITLE
#3909 Add GliderLink support

### PR DESCRIPTION
**Pull request by wargoth**

### CODE REVIEW

File: android/src/GliderLinkReceiver.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    this.context = context;
+
+    handler.post(new Runnable() {
+      @Override
+      public void run() {
+        context.registerReceiver(GliderLinkReceiver.this, new IntentFilter(GliderLinkReceiver.ACTION));
+      }
+    });
+  }
+
```
Comment (MaxKellermann) [inline android/src/GliderLinkReceiver.java]: There is lots of trailing whitespace in this file.

File: src/Android/GliderLink.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+GliderLink::Deinitialise(JNIEnv *env)
+{
+  gl_cls.Clear(env);
+}
+
+GliderLink* GliderLink::create(JNIEnv* env, Context* context,
+                                         unsigned index) {
+  assert(gl_cls != nullptr);
+
+  // Construct InternalGPS object.
```
Comment (MaxKellermann) [inline src/Android/GliderLink.cpp]: InternalGPS, wtf?

File: src/Android/GliderLink.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "GliderLink.hpp"
+#include "Util/StaticString.hxx"
```
Comment (MaxKellermann) [inline src/Android/GliderLink.cpp]: Why?

File: src/Android/GliderLink.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                              Angle::Degrees(latitude));
+
+  traffic->altitude_available = altitude > -10000.0;
+  traffic->altitude = altitude;
+  traffic->speed_received = gspeed >= 0.1;
+  traffic->speed = gspeed;
+  traffic->climb_rate_received = vspeed > -8675309.0;
+  traffic->climb_rate = vspeed;
+  traffic->track_received = bearing < 361.0;
+  traffic->track = Angle::Degrees(bearing);
```
Comment (MaxKellermann) [inline src/Android/GliderLink.cpp]: Don't assign this value if you know it's bad.

File: src/Android/GliderLink.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+  traffic->location_available = true;
+  traffic->location = GeoPoint(Angle::Degrees(longitude),
+                              Angle::Degrees(latitude));
+
+  traffic->altitude_available = altitude > -10000.0;
+  traffic->altitude = altitude;
+  traffic->speed_received = gspeed >= 0.1;
+  traffic->speed = gspeed;
+  traffic->climb_rate_received = vspeed > -8675309.0;
```
Comment (MaxKellermann) [inline src/Android/GliderLink.cpp]: What kind of magic number is this, and where is its documentation?

File: src/Android/GliderLink.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  traffic->name.SetASCII(nativeString);
+  env->ReleaseStringUTFChars(callsign, nativeString);
+
+  traffic->location_available = true;
+  traffic->location = GeoPoint(Angle::Degrees(longitude),
+                              Angle::Degrees(latitude));
+
+  traffic->altitude_available = altitude > -10000.0;
+  traffic->altitude = altitude;
+  traffic->speed_received = gspeed >= 0.1;
```
Comment (MaxKellermann) [inline src/Android/GliderLink.cpp]: Why `0.1`?

File: src/Android/GliderLink.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+  const char *nativeString = env->GetStringUTFChars(callsign, JNI_FALSE);
+  traffic->name.SetASCII(nativeString);
+  env->ReleaseStringUTFChars(callsign, nativeString);
+
+  traffic->location_available = true;
+  traffic->location = GeoPoint(Angle::Degrees(longitude),
+                              Angle::Degrees(latitude));
+
+  traffic->altitude_available = altitude > -10000.0;
```
Comment (MaxKellermann) [inline src/Android/GliderLink.cpp]: Why `-10000`?

File: src/Android/GliderLink.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  traffic->altitude = altitude;
+  traffic->speed_received = gspeed >= 0.1;
+  traffic->speed = gspeed;
+  traffic->climb_rate_received = vspeed > -8675309.0;
+  traffic->climb_rate = vspeed;
+  traffic->track_received = bearing < 361.0;
```
Comment (MaxKellermann) [inline src/Android/GliderLink.cpp]: So this allows a track of `360.999999` degrees. Is this what you wanted? Why does it allow a track of `-99999999` degrees?

File: src/Android/GliderLink.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    traffic_list.new_traffic.Update(basic.clock);
+  }
+
+  const char *nativeString = env->GetStringUTFChars(callsign, JNI_FALSE);
+  traffic->name.SetASCII(nativeString);
+  env->ReleaseStringUTFChars(callsign, nativeString);
+
+  traffic->location_available = true;
+  traffic->location = GeoPoint(Angle::Degrees(longitude),
```
Comment (MaxKellermann) [inline src/Android/GliderLink.cpp]: Location should be validated.

File: src/Device/Config.cpp [hunk 10]
```diff
@@ -242,6 +244,9 @@ DeviceConfig::GetPortName(TCHAR *buffer, size_t max_size) const
   case PortType::INTERNAL:
     return _("Built-in GPS & sensors");
 
+  case PortType::GLIDER_LINK:
+    return _("GliderLink traffic receiver");
```
Comment (MaxKellermann) [inline src/Device/Config.cpp]: Is it called "GliderLink receiver" or "GliderLink traffic receiver"?

File: src/Dialogs/Device/DeviceEditWidget.cpp [hunk 11]
```diff
@@ -68,6 +68,7 @@ static constexpr struct {
 #ifdef ANDROID
   { DeviceConfig::PortType::RFCOMM_SERVER, N_("Bluetooth server") },
   { DeviceConfig::PortType::DROIDSOAR_V2, _T("DroidSoar V2") },
+  { DeviceConfig::PortType::GLIDER_LINK, _T("GliderLink receiver") },
```
Comment (MaxKellermann) [inline src/Dialogs/Device/DeviceEditWidget.cpp]: Is it called "GliderLink receiver" or "GliderLink traffic receiver"? (see above)

File: src/GliderLink/GliderLinkData.hpp [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ }
+ */
+#ifndef XCSOAR_GLIDER_LINK_DATA_H
+#define XCSOAR_GLIDER_LINK_DATA_H
+
+#include "GliderLink/List.hpp"
+
+#include <type_traits>
```
Comment (MaxKellermann) [inline src/GliderLink/GliderLinkData.hpp]: Why?

File: src/GliderLink/GliderLinkId.hpp [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+ }
+ */
+
+#ifndef XCSOAR_GLIDER_LINK_ID_HPP
+#define XCSOAR_GLIDER_LINK_ID_HPP
+
+#include <stdint.h>
+
+/**
+ * The identification number of a GLIDER_LINK traffic.
```
Comment (MaxKellermann) [inline src/GliderLink/GliderLinkId.hpp]: What is `GLIDER_LINK`? Your nomenclature is somehwat inconsistent.

File: src/GliderLink/Traffic.hpp [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+#include "Rough/RoughSpeed.hpp"
+#include "Rough/RoughAngle.hpp"
+
+#include <type_traits>
+
+#include <tchar.h>
+
+struct GliderLinkTraffic {
+  /** Was the geographical location of the target received? */
+  bool location_available;
```
Comment (MaxKellermann) [inline src/GliderLink/Traffic.hpp]: Why does this flag exist when it's always true?

File: src/GliderLink/Traffic.hpp [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+#include "Geo/GeoPoint.hpp"
+#include "NMEA/Validity.hpp"
+#include "Util/StaticString.hxx"
+#include "Rough/RoughAltitude.hpp"
+#include "Rough/RoughSpeed.hpp"
+#include "Rough/RoughAngle.hpp"
+
+#include <type_traits>
+
+#include <tchar.h>
```
Comment (MaxKellermann) [inline src/GliderLink/Traffic.hpp]: Why?

File: src/GliderLink/Traffic.cpp [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+ You should have received a copy of the GNU General Public License
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ }
+ */
+
+#include "GliderLink/Traffic.hpp"
+
+void
+GliderLinkTraffic::Update(const GliderLinkTraffic &other)
```
Comment (MaxKellermann) [inline src/GliderLink/Traffic.cpp]: What is this undocumented method supposed to do, and who should call it?

File: src/GliderLink/Traffic.cpp [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ }
+ */
+
+#include "GliderLink/Traffic.hpp"
```
Comment (MaxKellermann) [inline src/GliderLink/Traffic.cpp]: Relative include path.

File: src/GliderLink/Traffic.hpp [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+struct GliderLinkTraffic {
+  /** Was the geographical location of the target received? */
+  bool location_available;
+
+  /** Was the direction of the target received? */
+  bool track_received;
```
Comment (MaxKellermann) [inline src/GliderLink/Traffic.hpp]: Why are some attributes called `*_available` and others `*_received`?

File: src/MapWindow/MapWindowTraffic.cpp [hunk 19]
```diff
@@ -116,6 +116,93 @@ MapWindow::DrawFLARMTraffic(Canvas &canvas,
...
+    // Points for the screen coordinates for the icon, name and average climb
+    PixelPoint sc, sc_name, sc_av, sc_alt;
+
+    // If FLARM target not on the screen, move to the next one
+    if (!projection.GeoToScreenIfVisible(target_loc, sc))
+      continue;
+
+    // Draw the callsign above the icon
+    sc_name = sc;
+    sc_name.x -= Layout::Scale(10);
```
Comment (MaxKellermann) [inline src/MapWindow/MapWindowTraffic.cpp]: Does this scale properly with all font sizes? Your formula doesn't consider font size at all.

File: src/GliderLink/Traffic.hpp [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+#include "Util/StaticString.hxx"
+#include "Rough/RoughAltitude.hpp"
+#include "Rough/RoughSpeed.hpp"
+#include "Rough/RoughAngle.hpp"
+
+#include <type_traits>
+
+#include <tchar.h>
+
+struct GliderLinkTraffic {
```
Comment (MaxKellermann) [inline src/GliderLink/Traffic.hpp]: This ordering of the struct attributes is bad for padding.

File: android/src/GliderLinkReceiver.java [hunk 21]
```diff
@@ -0,0 +1,116 @@
+/* Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2015 The XCSoar Project
```
Comment (MaxKellermann) [inline android/src/GliderLinkReceiver.java]: 2018, not 2015!

**Comment by MaxKellermann** [comment]
Your PR was not rebased on current master; it was a week old or so. Please keep it rebased; at least rebase right before you push.

**Comment by MaxKellermann** [comment]
Since the feature is available on Android only, it would be nice to save the overhead on all other targets and add an `#ifdef ANDROID` around the NMEAInfo attribute.

File: src/MapWindow/MapWindowTraffic.cpp (hunk 19 above)
Comment (wargoth) [reply src/MapWindow/MapWindowTraffic.cpp]: The implementation has been copied from FLARM

**Comment by wargoth** [comment]
> Since the feature is available on Android only, it would be nice to save the overhead on all other targets and add an `#ifdef ANDROID` around the NMEAInfo attribute.

I'm not sure how to test this, so I'll keep it simple

**Comment by wargoth** [comment]
Hi Max, any updates on this? Please let me know if there is anything I can help with.

**Comment by lordfolken** [comment]
Have a look at the CI log. It currently fails on the UNIX target, because it includes a java header file which is not available on UNIX. You need to add the #ifdef ANDROID around the code, so it doesn't get compiled into other targets at all.

**Comment by MaxKellermann** [comment]
Let's reopen this one as soon as the submitter fixes the build failures.

**Comment by wargoth** [comment]
Please reopen the pull request. I pushed more changes.

**Comment by lordfolken** [comment]
I think you need to open a new one as there is a force push.

**Pull request closed without merging**